Allow get_table_comment to fetch view comments too

necaris opened the pull request:
Since SQLAlchemy recommends that views be reflected in the same ways as tables (see https://docs.sqlalchemy.org/en/13/core/reflection.html#reflecting-views), allow the `get_table_comment` method to return metadata for views too.

keller00 commented on snowdialect.py:
```diff
@@ -583,7 +583,21 @@ def get_table_comment(self, connection, table_name, schema=None, **kw):
+                    (' IN SCHEMA {0}'.format(self.normalize_name(schema))) if schema else ''
+                )
+            cursor = connection.execute(sql_command)
+            ans = cursor.fetchone()
```
Could you make this a separate function, like `get_view_comment`?

keller00 replied:
Please install `tox` and make sure `tox -e fix_lint` passes.

keller00 replied:
Actually, don't worry about the `fix_lint`, it's made to work on our internal repo. I'll deal with linting issues.

keller00 commented on snowdialect.py:
```diff
@@ -583,7 +583,21 @@ def get_table_comment(self, connection, table_name, schema=None, **kw):
+        # Also, we should not throw an error here just because we could not
+        # fetch the *comment* field
+        comment = ans.get('comment', None) if ans else None
+        return {'text': comment if comment else None}
```
This if statement is not necessary. I'm pretty sure you catch all corner cases on the previous line.

keller00 replied:
You resolved this before I finished commenting 😄

akdor1154 commented:
ah I was just gonna PR for exactly this.. @necaris do you have time to update this to meet requirements or would you like me to pick this up?

necaris commented:
I'd have time to fix this up, what does it need?

On Tue, Feb 18, 2020, 22:34 Jarrad Whitaker <notifications@github.com>
wrote:

> ah I was just gonna PR for exactly this.. @necaris
> <https://github.com/necaris> do you have time to update this to meet
> requirements or would you like me to pick this up?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/snowflakedb/snowflake-sqlalchemy/pull/154?email_source=notifications&email_token=AADQPYLQOSETKHO2QUCV4YLRDSSD7A5CNFSM4KSRCCXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMGHGSI#issuecomment-588018505>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQPYLNMPDLUGZZW4HBBCLRDSSD7ANCNFSM4KSRCCXA>
> .
>

keller00 commented:
You could add a test, like how I have outlined and I also asked for your input on column comments.

necaris commented:
Thanks @keller00 I've been away from a keyboard this morning but I should be able to get to the points you mentioned tonight -- thanks for the review!

necaris commented:
@keller00 please note I've updated the PR and believe I've addressed all your comments, would appreciate another look. Note that I've factored `_get_view_comment` into a helper of its own, but since SQLAlchemy reflection doesn't have a distinct concept of reflecting views,  `get_table_comment` still needs to manage it.

Regarding column comments: since both `get_schema_columns` and `get_table_columns` use the [columns view](https://docs.snowflake.net/manuals/sql-reference/info-schema/columns.html) and the doc for that says that filtering on `table_name` also allows a view name to be used, I'd hope it's not a problem. However, I'm not sure how filtering by `table_schema` behaves in that view, so I don't know for sure.

keller00 commented on snowdialect.py:
```diff
@@ -572,18 +572,49 @@ def get_schema_names(self, connection, **kw):
-        ans = cursor.fetchone()
-        return {'text': ans['comment'] if ans['comment'] else None}
+        result = cursor.fetchone()
+        return result['comment']
```
This needs to be changed back.
There was another bug where the empty string comment would trick alembic into thinking that there's a comment.
We need to check `result['comment'] == ''` and then instead return `None`.

necaris replied:
@keller00 please note that I'm checking this value in `get_table_comment` now, rather than repeating the logic in both helper functions, so this should be taken care of.

keller00 commented on snowdialect.py:
```diff
@@ -572,18 +572,49 @@ def get_schema_names(self, connection, **kw):
+            )
+        cursor = connection.execute(sql_command)
+        result = cursor.fetchone()
+        return result['comment']
```
Same needs to be done here

necaris replied:
@keller00 please note that I'm checking this value in `get_table_comment` now, rather than repeating the logic in both helper functions, so this should be taken care of.

keller00 commented on snowdialect.py:
```diff
@@ -572,18 +572,49 @@ def get_schema_names(self, connection, **kw):
+        """
+        Returns comment of view in a dictionary as described by SQLAlchemy spec.
+        """
+        sql_command = "SHOW /* sqlalchemy:get_view_comment */ " \
```
The function name is missing an underscore here

keller00 commented on snowdialect.py:
```diff
@@ -572,18 +572,49 @@ def get_schema_names(self, connection, **kw):
-        Returns comment of table in a dictionary as described by SQLAlchemy spec
+        Returns comment of table in a dictionary as described by SQLAlchemy spec.
         """
         sql_command = "SHOW /* sqlalchemy:get_table_comment */ " \
```
Please add the missing underscore to the function name here

keller00 commented on snowdialect.py:
```diff
@@ -572,18 +572,49 @@ def get_schema_names(self, connection, **kw):
+        # this is a table or a view, we have to handle both cases here.
+        try:
+            comment = self._get_table_comment()
+            if comment is None:
```
Note that this logic will need to be updated once you change the return value checks.

keller00 commented on snowdialect.py:
```diff
@@ -572,18 +572,49 @@ def get_schema_names(self, connection, **kw):
+        SQLAlchemy expects.
+        """
+        # NOTE: Since SQLAlchemy may not (in fact, typically does not) know if
+        # this is a table or a view, we have to handle both cases here.
```
This might be nitpicky, but I think this could move into the doc string right above it 😄

keller00 commented on test/test_core.py:
```diff
@@ -552,6 +552,42 @@ def test_view_definition(engine_testaccount, db_parameters):
+        assert inspector.get_table_comment(test_view_name) == {'text': comment_text}
+    finally:
+        engine_testaccount.execute(
+            "DROP TABLE IF EXISTS {0}".format(test_table_name))
```
I'm just curious why did you wrap the `sql` in text on line 571, but not here.

necaris replied:
@keller00 now that you mention it, that does look odd -- I had in fact just copied it from the function above. I'll update them both.

keller00 commented:
Hey @necaris thanks again!
There are just some small things and the reintroduction of a bug that is not obvious. Once you have made these changes I'll submit it for testing and we can get it merged soon after.

necaris commented:
@keller00 I think I've got all your comments, please have another look. Thank you!

necaris commented on snowdialect.py:
```diff
@@ -572,18 +572,49 @@ def get_schema_names(self, connection, **kw):
+        except (KeyError, TypeError):
+            # Also, we should not throw an error here just because we could not
+            # fetch the *comment* field
+            comment = None
```
@keller00 I've found that calling e.g. `result['comment']` on an empty `result` row throws an error that SQLAlchemy _doesn't catch_, so it actually just breaks reflection -- hence wanting to guard against errors here and just return `None`.

keller00 replied:
You're right, but then this whole PR won't work.
When trying to retrieve a VIew comment it will throw an exception in `_get_table_comment` and then this will catch it and not try to retrieve the view comment

keller00 replied:
In light of this, maybe `return result` in the internal functions instead and check in this outer function `if result is None:`.
But I'd still remove this error catching.

necaris commented on snowdialect.py:
```diff
@@ -572,18 +572,49 @@ def get_schema_names(self, connection, **kw):
+        """
+        try:
+            comment = self._get_table_comment()
+            if comment is None or comment == "":
```
Updated! @keller00 thank you for the catch

keller00 commented:
There were a few bugs, but I have fixed them and will be accepting this PR.
Watch out for this in the next release.
Thank you @necaris 🏅 !

akdor1154 commented:
thanks @necaris !

necaris commented:
Thank you both for the help!

On Tue, Feb 25, 2020, 18:27 Jarrad Whitaker <notifications@github.com>
wrote:

> thanks @necaris <https://github.com/necaris> !
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/snowflakedb/snowflake-sqlalchemy/pull/154?email_source=notifications&email_token=AADQPYIMIA7A7EW3TP3Z6XTREWSPXA5CNFSM4KSRCCXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEM572PI#issuecomment-591134013>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQPYKMXA4JPTK2XYBVZXLREWSPXANCNFSM4KSRCCXA>
> .
>
